faunadb-scala/src/main/scala/faunadb/values/Result.scala

```diff
@@ -0,0 +1,69 @@
+package faunadb.values
+
+object Result {
+  def successful[T](t: T, path: FieldPath): Result[T] = VSuccess(t, path)
+
+  def collectErrors(rs: Result[_]*) =
+    VFail(rs.toList flatMap {
+      case VFail(errs) => errs
+      case _ => Nil
+    })
+
+  def NotFound[T](path: FieldPath): Result[T] =
+    VFail(List(FieldError("Value not found", path)))
+
+  def Unexpected[T](actual: Value, expected: String, path: FieldPath): Result[T] = {
+    val tpe = actual.getClass.getName
+    VFail(List(FieldError(s"Expected $expected value; found $tpe.", path)))
+  }
+
+  implicit class ValueResultOps(r: Result[Value]) {
+    def apply[T](field: Field[T]): Result[T] =
+      r match {
+        case VSuccess(v, pre) =>
+          field.get(v) match {
+            case VSuccess(t, path) => VSuccess(t, pre ++ path)
+            case VFail(errs) => VFail(errs map (_ prefixed pre))
+          }
+        case VFail(errs) => VFail(errs)
+      }
+
+    def apply(p: FieldPath, ps: FieldPath*): Result[Value] = apply(Field(p, ps: _*))
+
+    def as[T: Decoder]: Result[T] = apply(Field.as[T])
+
+    def collect[T](inner: Field[T]): Result[Seq[T]] = apply(Field.collect(inner))
+  }
+}
+
+sealed abstract class Result[+T] {
+
+  def get: T
+  def isDefined: Boolean
+
+  def map[U](f: T => U): Result[U]
+  def flatMap[U](f: T => Result[U]): Result[U]
+
+  def isEmpty = !isDefined
```

Comment:
It's the "None" case... results are never not satisfied, so it's not a separate axis like futures. I guess if it feels wrong, I could remove isDefined/isEmpty.